File: cpp/src/fences_demo.cpp
// Demonstrates std::atomic_thread_fence -- a standalone fence, not
// tied to any specific atomic variable, and the direct analog of Java
// 9's VarHandle.acquireFence()/releaseFence()/fullFence() (see
// java/FencesDemo.java).
//
// This is the "classic" way to build a publish/subscribe pattern using
// RELAXED atomics for the actual data, with the fences doing all the
// ordering work instead of attaching acquire/release directly to every
// individual load/store. Useful when you're publishing several values
// through one flag and don't want (or can't cheaply afford) full
// acquire/release semantics on each one individually -- the fence
// orders everything around it in program order, in one shot.

#include <atomic>
#include <thread>
#include <chrono>
#include <iostream>

std::atomic<int> payload{0};
std::atomic<int> morePayload{0};
std::atomic<bool> ready{false};

void produce() {
    payload.store(42, std::memory_order_relaxed);
    morePayload.store(43, std::memory_order_relaxed);
    std::atomic_thread_fence(std::memory_order_release); // orders BOTH stores above against...
    ready.store(true, std::memory_order_relaxed);          // ...this relaxed store
}

void consume() {
    while (!ready.load(std::memory_order_relaxed)) {       // relaxed load...
        std::this_thread::yield();
    }
    std::atomic_thread_fence(std::memory_order_acquire);   // ...paired with this fence
    std::cout << "payload = " << payload.load(std::memory_order_relaxed)
              << ", morePayload = " << morePayload.load(std::memory_order_relaxed) << "\n";
}

int main() {
    std::thread consumer(consume);
    std::this_thread::sleep_for(std::chrono::milliseconds(50)); // head start
    produce();
    consumer.join();

    // A standalone full (seq_cst) fence, with no particular variable
    // attached to it at all -- same idea as VarHandle.fullFence().
    std::atomic_thread_fence(std::memory_order_seq_cst);
    std::cout << "seq_cst fence executed: orders all prior loads/stores against all later ones\n";
    return 0;
}

File: java/src/FencesDemo.java
import java.lang.invoke.MethodHandles;
import java.lang.invoke.VarHandle;

/**
 * Since Java 9 (JEP 193), java.lang.invoke.VarHandle exposes the same
 * fine-grained access modes C++'s <atomic> gives you explicitly:
 *
 *   VarHandle mode      ~ C++ memory_order
 *   ---------------------------------------------
 *   plain get/set       ~ ordinary (non-atomic) access, no ordering
 *   getOpaque/setOpaque  ~ memory_order_relaxed  (atomicity, no ordering)
 *   getAcquire/setRelease~ memory_order_acquire / memory_order_release
 *   volatile get/set     ~ memory_order_seq_cst
 *
 * Before Java 9, `volatile` (always seq_cst-like) and the internal,
 * unsupported sun.misc.Unsafe class were the only ways to get any of
 * this in Java -- VarHandle is what finally brought a public,
 * standard API with an explicit ordering dial, about 5 years after
 * C++11 did the equivalent for C++.
 *
 * VarHandle ALSO exposes standalone fences, decoupled from any one
 * particular variable -- the direct analog of C++'s
 * std::atomic_thread_fence (see cpp/src/fences_demo.cpp):
 *
 *   VarHandle.acquireFence() / releaseFence() / fullFence()
 *   VarHandle.loadLoadFence() / storeStoreFence()
 *
 * A fence orders "everything around this point in program order,"
 * rather than being tied to the one memory access an acquire load or
 * release store carries its ordering on. That's useful when you're
 * publishing several plain/opaque values through a single flag and
 * don't want to (or can't) attach acquire/release semantics to every
 * individual field access.
 */
public class FencesDemo {

    static int payload;
    static int morePayload;
    static int ready; // plain field -- we publish it "by hand" via VarHandle, not volatile

    static final VarHandle READY;
    static {
        try {
            READY = MethodHandles.lookup()
                    .findStaticVarHandle(FencesDemo.class, "ready", int.class);
        } catch (ReflectiveOperationException e) {
            throw new ExceptionInInitializerError(e);
        }
    }

    static void produce() {
        payload = 42;
        morePayload = 43;
        VarHandle.releaseFence();       // orders BOTH plain writes above against...
        READY.setOpaque(1);              // ...this opaque (relaxed-atomicity) publish
    }

    static void consume() {
        while ((int) READY.getOpaque() == 0) { // opaque (relaxed-atomicity) read
            Thread.onSpinWait();
        }
        VarHandle.acquireFence();        // ...paired with this fence
        System.out.println("Consumer saw payload = " + payload + ", morePayload = " + morePayload);
    }

    public static void main(String[] args) throws InterruptedException {
        Thread consumer = new Thread(FencesDemo::consume);
        consumer.start();
        Thread.sleep(50); // give the consumer a head start spinning on `ready`
        produce();
        consumer.join();

        // A standalone full fence -- the seq_cst analog -- with no
        // particular variable attached to it at all:
        VarHandle.fullFence();
        System.out.println("fullFence() executed: orders all prior loads/stores against all later ones");
    }
}
